File: backend/app/schemas/supervision.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.models.supervision import ApprovalAction, ContentStatus, ContentType


class ContentBody(BaseModel):
    summary: str = ""
    questions: list[dict[str, Any]] = Field(default_factory=list)
    instructions: str = ""
    cards: list[dict[str, Any]] = Field(default_factory=list)
    sections: list[dict[str, Any]] = Field(default_factory=list)
    ai_confidence: float = 0.85
    ai_notes: str = ""


class AiContentCreate(BaseModel):
    course_id: str
    course_name: str
    title: str
    content_type: ContentType
    subject: str = "cs"
    concept_tags: list[str] = Field(default_factory=list)
    body: ContentBody


class AiContentUpdate(BaseModel):
    title: str | None = None
    body: ContentBody | None = None
    concept_tags: list[str] | None = None
    edit_note: str | None = None


class ReviewActionRequest(BaseModel):
    comment: str | None = None


class RejectRequest(BaseModel):
    reason: str = Field(min_length=3, max_length=1000)


class RegenerateRequest(BaseModel):
    instructions: str | None = None


class AiContentResponse(BaseModel):
    id: str
    course_id: str
    course_name: str
    title: str
    content_type: ContentType
    subject: str
    concept_tags: list[str]
    body: dict[str, Any]
    status: ContentStatus
    version: int
    teacher_id: str
    teacher_name: str
    reviewed_by_id: str | None
    reviewed_by_name: str | None
    rejection_reason: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RevisionResponse(BaseModel):
    id: str
    content_id: str
    version: int
    title: str
    body: dict[str, Any]
    edited_by_id: str
    edited_by_name: str
    edit_note: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ApprovalResponse(BaseModel):
    id: str
    content_id: str
    action: ApprovalAction
    reviewer_id: str
    reviewer_name: str
    reviewer_role: str
    comment: str | None
    previous_status: str
    new_status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class AuditLogResponse(BaseModel):
    id: str
    actor_id: str
    actor_name: str
    actor_role: str
    action: str
    resource_type: str
    resource_id: str
    details: dict[str, Any]
    created_at: datetime

    model_config = {"from_attributes": True}


class ContentDetailResponse(AiContentResponse):
    revisions: list[RevisionResponse] = Field(default_factory=list)
    approvals: list[ApprovalResponse] = Field(default_factory=list)


class SupervisionStatsResponse(BaseModel):
    pending_review: int
    approved: int
    rejected: int
    total: int

File: src/app/types/supervision.ts
export type ContentType =
  | 'quiz'
  | 'assignment'
  | 'flashcard'
  | 'study_guide'
  | 'lecture_note'
  | 'concept_map';

export type ContentStatus = 'pending_review' | 'approved' | 'rejected' | 'draft';

export type ApprovalAction = 'approved' | 'rejected' | 'regenerated' | 'edited';

export interface ContentBody {
  summary?: string;
  questions?: Array<{ question: string; options?: string[]; answer?: string }>;
  instructions?: string;
  cards?: Array<{ front: string; back: string }>;
  sections?: Array<{ heading: string; content: string }>;
  ai_confidence?: number;
  ai_notes?: string;
}

export interface AiContent {
  id: string;
  course_id: string;
  course_name: string;
  title: string;
  content_type: ContentType;
  subject: string;
  concept_tags: string[];
  body: ContentBody;
  status: ContentStatus;
  version: number;
  teacher_id: string;
  teacher_name: string;
  reviewed_by_id: string | null;
  reviewed_by_name: string | null;
  rejection_reason: string | null;
  created_at: string;
  updated_at: string;
}

export interface ContentRevision {
  id: string;
  content_id: string;
  version: number;
  title: string;
  body: ContentBody;
  edited_by_id: string;
  edited_by_name: string;
  edit_note: string | null;
  created_at: string;
}

export interface ApprovalRecord {
  id: string;
  content_id: string;
  action: ApprovalAction;
  reviewer_id: string;
  reviewer_name: string;
  reviewer_role: string;
  comment: string | null;
  previous_status: string;
  new_status: string;
  created_at: string;
}

export interface AuditLogEntry {
  id: string;
  actor_id: string;
  actor_name: string;
  actor_role: string;
  action: string;
  resource_type: string;
  resource_id: string;
  details: Record<string, unknown>;
  created_at: string;
}

export interface ContentDetail extends AiContent {
  revisions: ContentRevision[];
  approvals: ApprovalRecord[];
}

export interface SupervisionStats {
  pending_review: number;
  approved: number;
  rejected: number;
  total: number;
}

export const contentTypeLabels: Record<ContentType, string> = {
  quiz: 'Quiz',
  assignment: 'Assignment',
  flashcard: 'Flashcards',
  study_guide: 'Study Guide',
  lecture_note: 'Lecture Notes',
  concept_map: 'Concept Map',
};

export const statusLabels: Record<ContentStatus, string> = {
  pending_review: 'Pending Review',
  approved: 'Approved',
  rejected: 'Rejected',
  draft: 'Draft',
};

export const statusColors: Record<ContentStatus, string> = {
  pending_review: 'bg-[#F59E0B]/10 text-[#F59E0B]',
  approved: 'bg-[#22C55E]/10 text-[#22C55E]',
  rejected: 'bg-[#EF4444]/10 text-[#EF4444]',
  draft: 'bg-muted text-muted-foreground',
};
